### PR TITLE
dependencies: initial external dependency policy.

### PR DESCRIPTION
**Pull request by htuch**
This will apply to all changes to external dependencies in future PRs.

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

**Comment by htuch** [comment]
CC @envoyproxy/maintainers @envoyproxy/security-team

File: DEPENDENCY_POLICY.md
Comment (alyssawilk) [inline DEPENDENCY_POLICY.md:58]: IMO you can remove OSS here

File: DEPENDENCY_POLICY.md
Comment (alyssawilk) [inline DEPENDENCY_POLICY.md:81]: We really only update deps when needed for security reasons correct?  Generally if someone wants an upgrade due to some shiny new feature they have done the work and we review.  Are you suggesting we change that, or think it's worth clarifying?

File: tools/protodoc/requirements.txt
Comment (alyssawilk) [inline tools/protodoc/requirements.txt:1]: ?

**Review by alyssawilk — COMMENTED** [review]
This is awesome - thanks for putting it together!

File: EXTENSION_POLICY.md
Comment (moderation) [inline EXTENSION_POLICY.md:40]: Add link to generated External dependencies documentation?

## External dependencies report

The generated report is located at https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/external_deps

File: DEPENDENCY_POLICY.md
Comment (moderation) [inline DEPENDENCY_POLICY.md:45]: We should we make choosing a developer provided tarball over a Github generated tarball policy IMO - https://github.com/envoyproxy/envoy/blob/master/bazel/EXTERNAL_DEPS.md#choosing-tarballs

File: DEPENDENCY_POLICY.md
Comment (antoniovicente) [inline DEPENDENCY_POLICY.md:79]: It would be good to work our way towards updating most deps at least once per release if a new version of the dependency is available.  It would be fine to make such a change in a followup policy change.

File: DEPENDENCY_POLICY.md
Comment (htuch) [reply DEPENDENCY_POLICY.md:81]: I think we usually update dependencies continuously; @moderation has been responsible for a lot of this, see history at https://github.com/envoyproxy/envoy/commits/master/bazel/repository_locations.bzl.

There has been some high profile upgrades that were not really plausible to do mechanically as they required code changes, e.g. `http-parser`, but these are more the exception than rule.

I reckon it's better to try and stick with latest release version. The rationale for this is:
* Many code improvements are made in projects that can impact security that fall below the threshold of CVE.
* Many projects don't have long-term support branches, e.g. if we were talking Linux kernel, maybe it's fine to live on 4.0.x as you know this will have active support for a significant period of time.
* For folks who operate monorepos (cough), other projects force dependencies to be brought forward in time.

The downside is we have more churn and less burn time for a given dependency version.

File: tools/protodoc/requirements.txt
Comment (htuch) [reply tools/protodoc/requirements.txt:1]: Reverting, this was an experiment around https://github.com/envoyproxy/envoy/issues/12951

File: DEPENDENCY_POLICY.md
Comment (htuch) [reply DEPENDENCY_POLICY.md:45]: I assume you mean release tar ball, rather than generated SHA tarball, clarified.

File: DEPENDENCY_POLICY.md
Comment (htuch) [reply DEPENDENCY_POLICY.md:86]: I've left a comment aiming to track latest release version as an aspirational statement, which should be strictly stronger than this property.

Comment (antoniovicente) [reply DEPENDENCY_POLICY.md:86]: Thanks.

File: DEPENDENCY_POLICY.md
Comment (mattklein123) [inline DEPENDENCY_POLICY.md:42]: Can you detail the comment format? This comes up pretty often (I assume date imported, reason(s) not using a release version, etc.). cc @moderation

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks for putting this together. One small comment and I would suggest shipping and iterating.

File: DEPENDENCY_POLICY.md
Comment (htuch) [reply DEPENDENCY_POLICY.md]: Added. I actually plan on introducing a `last_updated` metadata field later, but will leave in the comment for now.

Comment (moderation) [reply DEPENDENCY_POLICY.md]: @htuch 's last update clarifying date format looks good. To really get some standardization I'd dictate that the date comment comes on the line after `project_url`. An example of non-conforming is https://github.com/envoyproxy/envoy/blob/master/bazel/repository_locations.bzl#L123. As soon as my pip3 PR merges I have a dataplane PR and will clean these up.

I like the idea of `last_updated` metadata (but probably only for the non-release deps)

Comment (htuch) [reply DEPENDENCY_POLICY.md]: For standardizing and consistency, I recommend holding up until we fix the last updated metadata, since that's the long term direction for sure.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**
